modules/core/src/smithy4s/internals/KeyDecoder.scala

```diff
@@ -0,0 +1,166 @@
+/*
+ *  Copyright 2021-2022 Disney Streaming
+ *
+ *  Licensed under the Tomorrow Open Source Technology License, Version 1.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *     https://disneystreaming.github.io/TOST-1.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package smithy4s.internals
+
+import java.util.Base64
+import java.util.UUID
+
+import smithy4s._
+import smithy4s.Document._
+import smithy4s.schema.EnumValue
+import smithy4s.schema.Primitive
+import smithy4s.schema.Primitive._
+import smithy4s.schema.SchemaVisitor
+
+trait KeyDecoder[A] { self =>
```

Comment:
**suggestion:** the name/package should probably mention that it's for document decoding

Reply:
I was not sure because I would like to use it for the SchemaVisitorJCodec

Reply:
hmm, it deals with the Document type directly though

Reply:
you're absolutely right, I'll need to rethink that one, my bad

I guess I'll adjust if I need to expand to support the JCodec thingy, or rename if I don't

Reply:
> I was not sure because I would like to use it for the SchemaVisitorJCodec 

Let's rename. The JCodec thing doesn't need it as much. What this solves at the document level is that the logic was very hack-ish, encoding the value as a Document before ensuring it was a String. 

The JCodec construct does not suffer from the same ugliness. We could refactor into a `isSimpleShape` function that'd take a Schema and output whether it's a primitive, or a transformation targeting a primitive, which would allow us to remove the `canBeKey` method on JCodec, but it's much less important.